src/satisfia/agents/makeMDPAgentSatisfia.py

```diff
@@ -241,6 +244,133 @@ def messingPotential_state(self, state): # recursive
+				aleph2target = interpolate(estAleph1, 2.0, aleph4state)
+				# Due to the new target aleph, we have to recompute the estimated alephs and resulting losses and propensities:
+				estAlephs2 = [estAspiration4action(state, actions[index], aleph2target) for index in indices]
+				propensities2 = [propensity(index, indices2, estAlephs2) for index in indices2]
+
+				if debug:
+					print("| localPolicyData", prettyState(state), aleph4state, {"a1": a1, "midTarget": midTarget, "estAleph1": estAleph1, "mid1": mid1, "indices2": indices2, "aleph2target": aleph2target, "estAlephs2": estAlephs2, "propensities2": propensities2})
+
+				# Like for a1, we now draw a2 using a softmin mixture of these actions, based on the new propentities,
+				# and get its admissibility interval:
+				i2 = distribution.categorical(indices2, propensities2).sample()
+				a2 = actions[i2]
+				adm2Lo, adm2Hi = adm2 = admissibility4action(state, a2)
+
+				# Now we need to find two aspiration intervals aleph1 in adm1 and aleph2 in adm2, 
+				# and a probability p such that
+				#   aleph1:p:aleph2 is contained in aleph4state 
+				# and aleph1, aleph2 are close to the estimates we used above in estimating loss.
+				# Instead of optimizing this, we use the following heuristic:
+				# We first choose p so that the midpoints mix exactly:
+				estAleph2 = estAlephs2[i2]
+				mid2 = midpoint(estAleph2)
+				p = relativePosition(mid1, midTarget, mid2)
+
+				# Now we find the largest relative size of aleph1 and aleph2 
+				# so that their mixture is still contained in aleph4state:
+				# we want aleph1Lo:p:aleph2Lo >= alephLo and aleph1Hi:p:aleph2Hi <= alephHi
+				# where aleph1Lo = mid1 - x * w1, aleph1Hi = mid1 + x * w1, 
+				#		 aleph2Lo = mid2 - x * w2, aleph2Hi = mid2 + x * w2,
+				# hence midTarget - x * w1:p:w2 >= alephLo and midTarget + x * w1:p:w2 <= alephHi,
+				# i.e., x <= (midTarget - alephLo) / (w1:p:w2) and x <= (alephHi - midTarget) / (w1:p:w2): 
+				w1 = estAleph1[1] - estAleph1[0]
+				w2 = estAleph2[1] - estAleph2[0]
+				w = interpolate(w1, p, w2)
+				x = min((midTarget - alephLo) / w, (alephHi - midTarget) / w) if (w > 0) else 0
+				aleph1 = [mid1 - x * w1, mid1 + x * w1]
+				aleph2 = [mid2 - x * w2, mid2 + x * w2]
+						
+				if debug:
+					print("| localPolicyData",prettyState(state), aleph4state, {"a1": a1, "estAleph1": estAleph1, "adm1": adm1, "w1": w1, "a2": a2, "estAleph2": estAleph2, "adm2": adm2, "w2": w2, "p": p, "w": w, "x": x, "aleph1": aleph1, "aleph2": aleph2})
+
+				if verbose or debug:
+					print(pad(state),"| | locPol, state",prettyState(state),"aleph4state",aleph4state,": a1,p,a2",a1,p,a2,"adm12",adm1,adm2,"aleph12",aleph1,aleph2)
+
+				return distribution.categorical([(a1, aleph1), (a2, aleph2)], [1-p, p]).sample()
+
+				# TODO: understand why in GW3 we go right with 50% probability rather than always going left.
+
+			# TODO: there is an optimization but it doesn't work yet
+		locPol = distribution.infer(sample)
```

Comment:
This sampling-based estimation needs to be replaced by an exhaustive-search-based exact computation. 

I.e., for each possible triple `triple = ((a1,aleph1), p, (a2,aleph2))`, we need to compute its probability `p_triple[triple]`. For this, the samplings in lines 296 and 327 need to be replaced by loops over all possible values of `i1` and `i2`. 

Then we construct the categorical distribution over pairs `(a,aleph)` by computing their effective probabilities as follows:
```
for ((a1,aleph1), p, (a2,aleph2)) in triples:
    p_effective[(a1,aleph1)] += 1-p
    p_effective[(a2,aleph2)] += p
```